### PR TITLE
Thread name improvements including coop conversion.

### PR DESCRIPTION
**Pull request by jaykrell**
- Convert setting thread name to be coop-compatible.

 - Set proper Windows kernel unicode thread names, not just the debugger ones.
  These don't require a debugger to be useful, can be retrieved, show up in ETW, etc.

 - For constant thread names, just retain the constant, not a copy.
   This includes producing constant unicode thread names on Windows.

 - For non-constant thread names, retain the UTF8, that it was going
   to produce anyway, instead of the unicode. While Unicode
   benefits Windows, retaining it does not.
   A downside here is that it moves around failure points.
   In particular, setting a managed thread name involved a memdup,
   now it involves utf16 to utf8 conversion, which hopefully succeeds.

 - Make setting a thread name often non-fatal, except where
   mandated by public API.

 - Switch two boolean parameters, that could use extension to three, to enums, for
   call site readability.

- Retain nul in thread name. It didn't matter before but is useful now.

- This does penalize ves_icall_System_Threading_Thread_GetName_internal, the one
  code path that benefited from retaining UTF16 here -- lossless/conversion-free passthrough from Set to Get, is nice.
  Arguably we should retain both.
  Notice that netcore removes this function.

- Factor copy/pasted code into new small function mono_gstring_append_thread_name.

- Remove unused fields from MonoInternalThread. But synchronize netcore and non-netcore size with the dream of netcore-ness being a command line switch.

Alternative to https://github.com/mono/mono/pull/15827.

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

File: mono/profiler/log.c [hunk 1]
```diff
@@ -3180,6 +3178,9 @@ profiler_thread_begin (const char *name, gboolean send)
 	return thread;
 }
 
+#define profiler_thread_begin(name, send)							\
+	profiler_thread_begin_function (name, MONO_WINDOWS_CONSTANT_THREAD_NAME (name), G_N_ELEMENTS (name), (send))
```
Comment (jaykrell) [inline mono/profiler/log.c]: Off by one here, fix after monolite attempt.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build Windows x64 C++

**Comment by jaykrell** [comment]
@monojenkins build Windows x64

**Comment by jaykrell** [comment]
@monojenkins build Windows x64 Interpreter

**Comment by jaykrell** [comment]
@monojenkins build Windows x64 Cooperative Suspend

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by jaykrell** [comment]
@monojenkins build Linux WebAssembly

**Comment by CoffeeFlux** [comment]
This is a somewhat involved PR - is there any chance of breaking it into smaller commits? I can review as-is but it's not ideal.

**Comment by jaykrell** [comment]
Somewhat. Some is inter-related. Some is motivated by other.

Possible plan:
 1 the unused fields -- https://github.com/mono/mono/pull/15910
 2 the windows kernel names -- can be done afterward; it is interrelated in that it motivates keeping the unicode data around briefly
 3 `mono_gstring_append_thread_name` can be done afterward I think, or factored out before and then changed later
 4 bool vs. flag -- can be done mostly separately/before (maybe shorter names?)
 5 factoring out `constant_no_error` names can be done separately/differently/before, and then massaged into new form.

but it isn't *that* big asis imho.

**Comment by jaykrell** [comment]
Try those 4 to start and then I'll whittle away again.

**Comment by jaykrell** [comment]
Merge conflicts piling up, good. :)

**Comment by jaykrell** [comment]
Substantially addresses https://github.com/mono/mono/issues/16248.

**Comment by lambdageek** [comment]
We took many parts of this (see PRs that link to this one).

**Comment by jaykrell** [comment]
Agreed, even without the PR open, I can go back and merge this with this master locally and see if anything is left, and PR that. Likely nothing.

**Pull request closed without merging**